README.md
# MAAE4907B_CDH
Capstone satellite design project for 22-23 C&amp;DH subsystem team. Testing


hal/changelog.txt
Changelog for ISIS OBC HAL library

Table of contents

Changelog for:
- version 0.2.4 (11 Dec 2015)
- version 0.2.3 (13 Nov 2015)
- version 0.2.2 (22 Sep 2015)
- Version 0.2.1 (23 Jun 2015)
- Version 0.2.0 (20 Apr 2015)
- Version 0.1.9 (07 Apr 2015)
- Version 0.1.8 (02 Mar 2015)
- Version 0.1.7 (26 Feb 2015)
- Version 0.1.6 (25 Feb 2015)
- Version 0.1.5 (27 Jan 2015)
- Version 0.1.4 (22 Dec 2014)
- Version 0.1.3 (10 Dec 2014)
- Version 0.1.2 (11 Nov 2014)
- Version 0.1.1b (24 Oct 2014)
- Version 0.1.0 (15 Oct 2014)

Version 0.2.4
==============

Summary: Concurrent file handle increase, I2C slave timeout improvement
Release date: 11th of December 2015
Prepared by: Jonathan Ward

Specific changes

* Increased the maximum number of concurrent filehandles that can simultaneously be open to 20

General changes

Drivers/I2Cslave.c
* Improved accuracy of the I2C timeout measurement

Version 0.2.3
==============

Summary: I2C stability improvements
Release date: 13th of November 2015
Prepared by: Jonathan Ward & Gerard Aalbers

Specific changes

Drivers/I2Cslave.c
* Improved I2C perhipheral status handling after I2C slave stress testing

Drivers/I2C.c
* Disabled interleaving of I2C master transfers since this introduced significant complexity and potential instability for very little gain
* Disabled transfer timeout for single byte reads to improve dealing with clock stretching by slaves

Timing/Time.h & Timing/Time.c
* Moved definition of Unix epoch at 01-01-2000 00:00:00 UTC to header file (this is the earliest time the system will accept)

Version 0.2.2
==============

Summary: FreeRTOS updates, interrupt based DBGU, filesystem changes
Release date: 22nd of September 2015
Prepared by: Pieter Botma

General changes

FreeRTOS
* Updated FreeRTOS from version 7.5.0 to version 7.5.3, please see http://www.freertos.org/History.txt for a detailed changelog
* Updated port of FreeRTOS. This fixes a context switching issue, which was visible as UART starvation and FRAM contention.

Specific changes

Utility/dgbu_int.h & Utility/dbgu_int.c
* Added new interrupt based method for acquiring input from debug console, which is more reliable than the polling version

FileSystem
* Changed method of selecting and initializing a specific SD card to now use f_initvolume( drvnumber, driver_init, driver_param ) in HCC API
* Maximum number of simultaneous tasks using the filesystem increased from 4 to 8

Demo/demo_sd.c
* SD demo has been updated to use new SD card selection and initialization method

Demo/demo_uart.c
* UART demo now properly tests starvation scenario

Version 0.2.1
==============

Summary: 64 MB SDRAM support, doxygen clarifications, various smaller fixes
Release date: 23rd of June 2015
Prepared by: Gerard Aalbers

Generic changes

* This release introduces support for iOBCs with 64 MB of SDRAM

* Added several clarifications to doxygen documentation (Storage/FRAM.h, supervisor.h, and AT91 - dbgu.h)

Specific changes

Demo/demo_sd.c
* Improved doxygen and code examples 

Drivers/UART.h & Drivers/UART.c
* Fixed issue with UART_setRxEnabled not working correctly for UART bus 2 
* Moved UART queue size definition to header file

Storage/FRAM.h & Storage/FRAM.c
* Fixed definition of FRAM write protect structure
* Enabled error code indicating FRAM write protect misconfiguration
* Fixed issue with last byte of FRAM was not being addressable

AT91 - pio.c
* Added checks for NULL pointers on user specified pins and pin lists

Version 0.2.0
==============

Summary: introduction of journalling filesystem, doxygen changes
Release date: 20th of April 2015
Prepared by: Gerard Aalbers & Pieter Botma

Generic changes

* This release introduces a reliable, journalling fileystem for the iOBC. This filesystem will increase the reliability of non-volatile storage on SD cards and eliminate the susceptibility of file or filesystem corruption due to unexpected power cycles during write operations. This release will likely introduce significant changes when using the SD card storage.
	- Please note that full reliability can only be achieved when using recommended SwissBit SD cards
	- The most significant changes and points of attention are:
		+ The introduction of safe and non_safe methods, most notably for opening a file. In general all operations on a safely opened file are slower.
		+ Slightly different function parameters compared to the previous filesystem
		+ Writes are not saved to disk until f_flush() or f_close() is called on that file
		+ Reduced performance when safely writing using f_flush() but standard performance when reading
		+ The mode in which a file is opened affects if multiple handles to that file are allowed
		+ The previous file system referred to volume id's as numbers (0, 1, etc), whereas the new filesystem refer to them as letters (a, b, etc)
		+ The new filesystem API closely resembles the API from <stdio.h> and care should be taken not to mix the calls and defines
	- Recommend starting points are the following files:
		+ \hal\docs\HCC FAT and SafeFAT File System User's Guide v3.50.pdf - user guide of the filesystem including API documentation
		+ \hal\hcc\include\hcc\api_fat.h - main interface file
		+ \hal\demo\src\Demo\demo_sd.c - contains demonstration code on how to use the new filesystem and its API

* Several doxygen documentation improvements and renamed directory with doxygen from html to doxygen for clarity

Specific changes

Storage\FRAM.c
* Fixed issue where calling FRAM functions would cause a reset if FRAM hadn't been initialized yet

Utility\fprintf.h & Utility\fprintf.c
* f_puts() and f_printf() functions (which use the new filesystem) can now be found here

Version 0.1.9
==============

Summary: checksum improvements, I2C slave fix
Release date: 7th of April 2015
Prepared by: Gerard Aalbers

Generic changes


hal-demo/src/Tests/RTCTest.h
/*
 * RTCTest.h
 *
 *  Created on: Mar. 24, 2022
 *      Author: Sam Dunthorne
 */

#ifndef TESTS_RTCTEST_H_
#define TESTS_RTCTEST_H_

Boolean RTCTest(void);

#endif /* TESTS_RTCTEST_H_ */


hal-demo/src/Tests/SPITest.h
/*
 * SPITest.h
 *
 *  Created on: Mar. 24, 2022
 *      Author: Sam Dunthorne
 */

#ifndef TESTS_SPITEST_H_
#define TESTS_SPITEST_H_

Boolean SPITest(void);

#endif /* TESTS_SPITEST_H_ */


hal-demo/src/Tests/cspTest.c
/*
 * cspTest.cpp
 *
 *  Created on: Nov. 25, 2021
 *      Author: Tyler
 */

#include "cspTest.h"
#include <csp/drivers/i2c.h>

#define I2C_SPEED 			400000
#define I2C_TIMEOUT 		1000

/* Server port, the port the server listens on for incoming connections from the client. */
#define MY_SERVER_PORT		10

/* Commandline options */
static uint8_t server_address = 255;

/* test mode, used for verifying that host & client can exchange packets over the loopback interface */
static bool test_mode = false;
static unsigned int server_received = 0;

Boolean cspTest() {

	//Startup CSP Test
	uint8_t address = 1;
	csp_debug_level_t debug_level = CSP_DEBUG;

	//If NULL, default "0/0 I2C" expected
	const char * rtable = NULL;

	/* enable/disable debug levels */
	for (csp_debug_level_t i = 0; i <= CSP_LOCK; ++i) {
		csp_debug_set_level(i, (i <= debug_level) ? true : false);
	}

	csp_log_info("Initializing CSP");

	/* Init CSP with address and default settings */
	csp_conf_t csp_conf;
	csp_conf_get_defaults(&csp_conf);
	csp_conf.address = address;
	int error = csp_init(&csp_conf);
	if (error != CSP_ERR_NONE) {
		csp_log_error("csp_init() failed, error: %d", error);
		exit(1);
	}

	/* Start router task with 500 bytes of stack (priority is only supported on FreeRTOS) */
	csp_route_start_task(500, configMAX_PRIORITIES-2);

	/* Add interface(s) */
	csp_iface_t * default_iface = NULL;
	error = csp_I2C_start_and_add_I2C_interface(I2C_SPEED, I2C_TIMEOUT, CSP_IF_I2C_DEFAULT_NAME,  &default_iface);
	if (error != CSP_ERR_NONE) {
		csp_log_error("failed to add I2C interface, error: %d", error);
		exit(1);
	}


	if (rtable) {
		error = csp_rtable_load(rtable);
		if (error < 1) {
			csp_log_error("csp_rtable_load(%s) failed, error: %d", rtable, error);
			exit(1);
		}
	} else if (default_iface) {
		csp_rtable_set(CSP_DEFAULT_ROUTE, 0, default_iface, CSP_NO_VIA_ADDRESS); //WILL BE "0/0 I2C"
	} else {
		/* no interfaces configured - run server and client in process, using loopback interface */
		server_address = address;
	}

	printf("Connection table\r\n");
	csp_conn_print_table();

	printf("Interfaces\r\n");
	csp_route_print_interfaces();

	printf("Route table\r\n");
	csp_route_print_table();

	//TODO - Create a waiting loop here for test??

	/* Start server thread */
	if ((server_address == 255) ||  /* no server address specified, I must be server */
		(default_iface == NULL)) {  /* no interfaces specified -> run server & client via loopback */
		//csp_thread_create(task_server, "SERVER", 1000, NULL, 0, NULL);
	}

	/* Start client thread */
	if ((server_address != 255) ||  /* server address specified, I must be client */
		(default_iface == NULL)) {  /* no interfaces specified -> run server & client via loopback */
		//csp_thread_create(task_client, "CLIENT", 1000, NULL, 0, NULL);
	}

	/* Wait for execution to end (ctrl+c) */
	while(1) {
		csp_sleep_ms(3000);

		if (test_mode) {
			/* Test mode is intended for checking that host & client can exchange packets over loopback */
			if (server_received < 5) {
				csp_log_error("Server received %u packets", server_received);
				exit(1);
			}
			csp_log_info("Server received %u packets", server_received);
			exit(0);
		}
	}

	return 0;


    return TRUE;
}




hal-demo/src/cspSrc/drivers/i2c/i2c.c


#include <csp/drivers/i2c.h>

#include <stdio.h>

#include <csp/csp.h>
#include <hal/Drivers/I2C.h>
#include <csp/arch/csp_malloc.h>
#include <csp/interfaces/csp_if_i2c.h>
#include <csp/csp_debug.h>

typedef struct {
	char name[CSP_IFLIST_NAME_MAX + 1];
	csp_iface_t iface;
	csp_i2c_interface_data_t ifdata;
	unsigned int slaveAddress;

} i2c_context_t;

/* -for reference-
 * typedef struct _I2CwriteReadTransfer {
	unsigned int slaveAddress; //!< Address of the slave where to make the transfer.
	unsigned int writeSize; //!< Number of bytes to be written to the I2C slave.
	unsigned int readSize; //!< Number of bytes to be read from the I2C slave.
	unsigned char *writeData; //!< Memory location of the data to be written to the I2C slave.
	volatile unsigned char *readData; //!< Memory location to store the data read from the I2C slave.
	portTickType writeReadDelay; //!< A delay inserted between writing to an I2C slave and reading back from it.
} I2Ctransfer;
 */

//WIP
/**
 * FROM INTERFACE
 * ifroute->iface->driver_data (SELF REFERENCE - NOT USED), frame
 *
 * TO HAL DRIVER
 * unsigned int slaveAddress, unsigned char *data, unsigned int size
 *
 * @param driver_data
 * @param data
 * @param data_length
 * @return
 */
static int i2c_driver_tx(void *driver_data, csp_i2c_frame_t * frame) {

	i2c_context_t * ctx = driver_data;
	I2CtransferStatus txResult = I2C_write(frame->dest, frame->data, frame->len);
	if (txResult == 0) {
		csp_buffer_free(frame);
		return CSP_ERR_NONE;
	}
	csp_log_error("I2C Driver write error code: %d", txResult);
	return CSP_ERR_TX;
}

//TODO - Integrate our driver for incoming transmissions so CSP can read it
static void i2c_driver_rx(void * user_data, uint8_t * data, size_t data_size, void * pxTaskWoken) {

	//Note our under the hood pseudo multi master mode needs to be implemented such that this will work
	//kiss_context_t * ctx = user_data;
	//csp_i2c_rx(&ctx->iface, data, data_size, NULL);
}

int csp_I2C_start_and_add_I2C_interface(unsigned int i2cBusSpeed_Hz,unsigned int i2cTransferTimeout, const char * ifname, csp_iface_t ** return_iface) {

	if (ifname == NULL) {
		ifname = CSP_IF_I2C_DEFAULT_NAME;
	}

	csp_log_info("INIT %s: Speed: %d hz, Timeout: %d ",
			ifname, i2cBusSpeed_Hz, i2cTransferTimeout);

	i2c_context_t * ctx = csp_calloc(1, sizeof(*ctx));
	if (ctx == NULL) {
		return CSP_ERR_NOMEM;
	}

	strncpy(ctx->name, ifname, sizeof(ctx->name) - 1);
	ctx->iface.name = ctx->name;
	ctx->iface.driver_data = ctx;
	ctx->iface.interface_data = &ctx->ifdata;
	ctx->ifdata.tx_func = i2c_driver_tx;

	int res = csp_i2c_add_interface(&ctx->iface);
	if (res == CSP_ERR_NONE) {
		// Our HAL Driver startup - NOTE for multimaster support we may need to call a function in a custom wrapper that implements the MM flowchart
		res = I2C_start(i2cBusSpeed_Hz, i2cTransferTimeout);
		if(res != 0) {
			TRACE_FATAL("\n\r I2Ctest: I2C_start returned: %d! \n\r", res);
		}

		//This created RX Task needs to interact with our I2Cslave driver that operates via pseudo master mode flow chart wrapper solution
		xTaskHandle taskQueuedI2Ctest3Handle;
		csp_thread_create(csp_i2c_rx, "i2c_rx", 1024, NULL, configMAX_PRIORITIES-2, &taskQueuedI2Ctest3Handle);
		int retValInt = 0;
	}

	if (return_iface) {
		*return_iface = &ctx->iface;
	}

	return res;
}


hal-demo/src/main.c
/*
 * main.c

 *      Author: Akhil
 *      Alteration by C&DH Team
 */

#include "Demo/demo_sd.h"

#include "Tests/legacyTests/ADCtest.h"
#include "Tests/legacyTests/boardTest.h"
#include "Tests/legacyTests/checksumTest.h"
#include "Tests/legacyTests/FloatingPointTest.h"
#include "Tests/legacyTests/I2CslaveTest.h"
#include "Tests/legacyTests/I2Ctest.h"
#include "Tests/legacyTests/LEDtest.h"
#include "Tests/legacyTests/PinTest.h"
#include "Tests/legacyTests/PWMtest.h"
#include "Tests/legacyTests/SDCardTest.h"
#include "Tests/legacyTests/SPI_FRAM_RTCtest.h"
#include "Tests/legacyTests/SupervisorTest.h"
#include "Tests/legacyTests/TimeTest.h"
#include "Tests/legacyTests/UARTtest.h"
#include "Tests/legacyTests/USBdeviceTest.h"

#include "Tests/FreeRTOSTest.h"
#include "Tests/cspTest.h"
#include "Tests/I2CmultiMasterTest.h"
#include "Tests/TaskSequenceTimeTest.h"
#include "Tests/FRAMTest.h"
#include "Tests/SPITest.h"
#include "Tests/UARTReadTest.h"
#include "Tests/UARTReadWriteTest.h"
#include "Tests/UARTWriteTest.h"
#include "Tests/RTCTest.h"
#include "Tests/EPSTelemetryTest.h"

#include <at91/utility/exithandler.h>
#include <at91/commons.h>
#include <at91/utility/trace.h>
#include <at91/peripherals/cp15/cp15.h>

#include <freertos/FreeRTOS.h>
#include <freertos/task.h>
#include <freertos/semphr.h>

#include <hal/Timing/WatchDogTimer.h>
#include <hal/Drivers/LED.h>
#include <hal/Utility/util.h>

#include <hal/boolean.h>
#include <hal/version/version.h>

#include <hcc/api_fat_test.h>
#include <hcc/api_hcc_mem.h>
#include <hcc/api_fat.h>

#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#define ENABLE_MAIN_TRACES 1
#if ENABLE_MAIN_TRACES
	#define MAIN_TRACE_INFO			TRACE_INFO
	#define MAIN_TRACE_DEBUG		TRACE_DEBUG
	#define MAIN_TRACE_WARNING		TRACE_WARNING
	#define MAIN_TRACE_ERROR		TRACE_ERROR
	#define MAIN_TRACE_FATAL		TRACE_FATAL
#else
	#define MAIN_TRACE_INFO(...)	{ }
	#define MAIN_TRACE_DEBUG(...)	{ }
	#define MAIN_TRACE_WARNING(...)	{ }
	#define MAIN_TRACE_ERROR		TRACE_ERROR
	#define MAIN_TRACE_FATAL		TRACE_FATAL
#endif

Boolean selectAndExecuteTest() {
	unsigned int selection = 0;
	Boolean offerMoreTests = TRUE;

	printf( "\n\r Select a test to perform: \n\r");
	printf("\t 1) I2C \n\r");
	printf("\t 2) SD-Card File System \n\r");
	printf("\t 3) SPI + FRAM + RTC \n\r");
	printf("\t 4) UART \n\r");
	printf("\t 5) ADC Single Shot \n\r");
	printf("\t 6) ADC Continuous Mode \n\r");
	printf("\t 7) Aux Pins \n\r");
	printf("\t 8) LED \n\r");
	printf("\t 9) PWM \n\r");
	printf("\t 10) USB Device \n\r");
	printf("\t 11) Supervisor Controller Test - SPI interface \n\r");
	printf("\t 12) Supervisor Controller Test - I2C interface \n\r");
	printf("\t 13) Board Test \n\r");
	printf("\t 14) Time Test \n\r");
	printf("\t 15) Checksum Test \n\r");
	printf("----- CuSAT Additions 2021_2022 -----\n\r");
	printf("\t 16) I2C Slave Test \n\r");
	printf("\t 17) FreeRTOS Test \n\r");
	printf("\t 18) CSP-AX100 Test \n\r");
	printf("\t 19) MultiMaster Test \n\r");
	printf("\t 20) SPI Test \n\r");
	printf("\t 21) UART Read \n\r");
	printf("\t 22) UART Write \n\r");
	printf("\t 23) UART Read And Write \n\r");
	printf("\t 24) RTC Test \n\r");
	printf("\t 25) FRAM Test \n\r");
	printf("\t 26) EPSCmd \n\r");
	printf("\t 27) Task Sequence Time Test \n\r");

	while(UTIL_DbguGetIntegerMinMax(&selection, 1, 27) == 0); //Update the range if you add functions

	//Verify you are importing the header file for any new tests you add that are in a new file
	switch(selection) {
	case 1:
		offerMoreTests = I2Ctest();
		break;
	case 2:
		offerMoreTests = SDCardTest();
		break;
	case 3:
		offerMoreTests = SPI_FRAM_RTCtest();
		break;
	case 4:
		offerMoreTests = UARTtest();
		break;
	case 5:
		offerMoreTests = ADCtestSingleShot();
		break;
	case 6:
		offerMoreTests = ADCtest();
		break;
	case 7:
		offerMoreTests = PinTest();
		break;
	case 8:
		offerMoreTests = LEDtest();
		break;
	case 9:
		offerMoreTests = PWMtest();
		break;
	case 10:
		offerMoreTests = USBdeviceTest();
		break;
	case 11:
		offerMoreTests = SupervisorTest(FALSE);
		break;
	case 12:
		offerMoreTests = SupervisorTest(TRUE);
		break;
	case 13:
		offerMoreTests = boardTest();
		break;
	case 14:
		offerMoreTests = timeTest();
		break;
	case 15:
		offerMoreTests = checksumTest();
		break;
		//Our Tests Begin
	case 16:
		offerMoreTests = I2CslaveTest();
		break;
	case 17:
		offerMoreTests = freeRTOSTest();
		break;
	case 18:
		offerMoreTests = cspTest();
		break;
	case 19:
		offerMoreTests = I2CmultiMasterTest();
		break;
	case 20:
		offerMoreTests = SPITest();
		break;
	case 21:
		offerMoreTests = UARTReadTest();
		break;
	case 22:
		offerMoreTests = UARTWriteTest();
		break;
	case 23:
		offerMoreTests = UARTReadWriteTest();
		break;
	case 24:
		offerMoreTests = RTCTest();
		break;
	case 25:
		offerMoreTests = FRAMTest();
		break;
	case 26:
		offerMoreTests = EPSTelemetryTest();
		break;
	case 27:
		offerMoreTests = TaskSequenceTimeTest();
		break;
	default:
		break;
	}

	return offerMoreTests;
}

void taskMain() {
	unsigned int choice;
	Boolean offerMoreTests = FALSE;

	WDT_startWatchdogKickTask(10 / portTICK_RATE_MS, FALSE);

	while(1) {
		LED_toggle(led_1);

		offerMoreTests = selectAndExecuteTest();

		/*
		 * This section works by going out to a test function, this function will typically generate test tasks.
		 * If the test function returns false, we will fall out of the main menu and yield to the next task, allowing them to go.
		 * If the test function returns true you enter the block below and may run more tests, in effect you can stack them up and let them all go at once
		 * as soon as you answer NO for perform more tests, we fall out of the menu loop and yield to the other tasks
		 *
		 * In some cases we opt to not use tasks and instead run the test inline with the menu task and when we return to menu when the test is complete.
		 * In this way we can run a test over and over, good for I2C testing for example when we want to send many times without reboot.
		 *
		 */

		if(offerMoreTests != FALSE) {
			// Not all tests will actually exit, so we may not reach here.
			// Even when we do, its good to be careful not to simultaneously run too many tests.
			// Instead, reboot between the tests.
			// In some cases, you must select 0 here to allow the test-tasks to do their work.
			printf("Perform more tests? (1=Yes, 0=No): \n\r");
			// No WatchDog resets here: The DBGU driver resets the WatchDog while it waits for user to enter characters.
			while(UTIL_DbguGetIntegerMinMax(&choice, 0, 1) == 0);
			if(choice == 0) {
				break;
			}
		}
		else {
			break;
		}
	}

	// Suspend itself.
	//vTaskSuspend(NULL);

	while(1) {
		//We added the LED wave to indicate the main menu tasks is complete and yielding to other tasks
		LED_wave(1);
		LED_waveReverse(1);
		LED_wave(1);
		LED_waveReverse(1);
		vTaskDelay(500);
	}

}

int main() {
	unsigned int i = 0;
	xTaskHandle taskMainHandle;

	TRACE_CONFIGURE_ISP(DBGU_STANDARD, 2000000, BOARD_MCK);
	// Enable the Instruction cache of the ARM9 core. Keep the MMU and Data Cache disabled.
	CP15_Enable_I_Cache();

	printf("\n\r -- iobcBase Program Booted --\n\r");

	LED_start();

	// The actual watchdog is already started, this only initializes the watchdog-kick interface.
	WDT_start();

	printf("\nDemo applications for ISIS OBC Hardware Abstraction Layer Library built on %s at %s\n", __DATE__, __TIME__);
	printf("\nDemo applications use:\n");
	printf("* HAL lib version %s.%s.%s built on %s at %s\n", HalVersionMajor, HalVersionMinor, HalVersionRevision,
				HalCompileDate, HalCompileTime);

	LED_wave(1);
	LED_waveReverse(1);
	LED_wave(1);
	LED_waveReverse(1);

	MAIN_TRACE_DEBUG("\t main: Starting main task.. \n\r");
	xTaskGenericCreate(taskMain, (const signed char*)"taskMain", 4096, NULL, configMAX_PRIORITIES-2, &taskMainHandle, NULL, NULL);

	MAIN_TRACE_DEBUG("\t main: Starting scheduler.. \n\r");
	vTaskStartScheduler();

	// This part should never be reached.
	MAIN_TRACE_DEBUG("\t main: Waiting in an infinite loop. \n\r");
	while(1) {
		LED_wave(1);
		MAIN_TRACE_DEBUG("MAIN: STILL ALIVE %d\n\r", i);
		i++;
	}

	return 0;
}
